### DISCUSSION TITLE
Build as static binary?

### DISCUSSION THREAD

**Discussion by schildbach** [tinyproxy/tinyproxy#527] [Q&A] [answered]
Is it possible to build tinyproxy in an all-static way, so that it can run without OS (only kernel)? If so, how?

**Comment by rofl0r** [comment, marked as answer]
an OS is a kernel, basically. so your question doesnt make sense.
if you have a linux kernel, and a libc other than glibc, you can statically link and run tinyproxy just fine.
for example on my musl-based distro, i just need `LDFLAGS=-static ./configure ; make` and the tinyproxy binary produced will be completely independant of any installed libraries and work on any machine with a recent enough linux kernel and same CPU architecture.
if you want to run it on bare metal instead, there's a thing called "root kernel" which you can link into your executable, but i never tried that out.

**Comment by schildbach** [comment]
Thanks, that worked.

("OS" also contains all the dynamic libraries and much more, so what I meant was "run on an otherwise empty (read-only) filesystem". In Docker lingo, this would be `FROM scratch`.)

**Reply by rofl0r** [reply]
one file is always needed, that's /etc/resolv.conf. otherwise tinyproxy can't know which DNS server to use. speaking of DNS, that's also the reason you can't make a static binary on glibc, because their DNS resolver uses a plugin-based framework called NSS which uses dlopen and requires to load some shared objects even if the program was static linked.
